src/sentry/sentry_metrics/indexer/ratelimiters.py

```diff
@@ -0,0 +1,151 @@
+    }[use_case_id]
+
+    if org_id is not None:
+        return f"metrics-indexer-{use_case_str}-org-{org_id}"
+    else:
+        return f"metrics-indexer-{use_case_str}-global"
+
+
+@metrics.wraps("sentry_metrics.indexer.construct_quotas")
+def _construct_quotas(use_case_id: UseCaseKey) -> Sequence[Quota]:
+    """
+    Construct write limit's quotas based on current sentry options.
+
+    This value can potentially cached globally as long as it is invalidated
+    when sentry.options are.
+    """
+    if use_case_id == UseCaseKey.PERFORMANCE:
+        return [
+            Quota(prefix_override=_build_quota_key(UseCaseKey.PERFORMANCE, None), **args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.performance.global")
+        ] + [
+            Quota(prefix_override=None, **args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.performance.per-org")
+        ]
+    elif use_case_id == UseCaseKey.RELEASE_HEALTH:
+        return [
+            Quota(prefix_override=_build_quota_key(UseCaseKey.RELEASE_HEALTH, None), **args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.releasehealth.global")
+        ] + [
+            Quota(prefix_override=None, **args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.releasehealth.per-org")
+        ]
+    else:
+        raise ValueError(use_case_id)
+
+
+@metrics.wraps("sentry_metrics.indexer.construct_quota_requests")
+def _construct_quota_requests(
+    use_case_id: UseCaseKey, keys: KeyCollection
+) -> Tuple[Sequence[OrgId], Sequence[RequestedQuota]]:
+    org_ids = []
+    requests = []
+    quotas = _construct_quotas(use_case_id)
+
+    if quotas:
+        for org_id, strings in keys.mapping.items():
+            org_ids.append(org_id)
+            requests.append(
+                RequestedQuota(
+                    prefix=_build_quota_key(use_case_id, org_id),
+                    requested=len(strings),
+                    quotas=quotas,
+                )
+            )
+
+    return org_ids, requests
+
+
+RateLimitState = Tuple[UseCaseKey, Sequence[RequestedQuota], Sequence[GrantedQuota], Timestamp]
+
+
+class WritesLimiter:
+    def __init__(self) -> None:
+        self.rate_limiters: MutableMapping[UseCaseKey, RedisSlidingWindowRateLimiter] = {}
+
+    def _get_rate_limiter(self, use_case_id: UseCaseKey) -> RedisSlidingWindowRateLimiter:
+        if use_case_id not in self.rate_limiters:
+            options = get_ingest_config(use_case_id).writes_limiter_cluster_options
+            self.rate_limiters[use_case_id] = RedisSlidingWindowRateLimiter(**options)
+
+        return self.rate_limiters[use_case_id]
+
+    @metrics.wraps("sentry_metrics.indexer.check_write_limits")
+    def check_write_limits(
+        self, use_case_id: UseCaseKey, keys: KeyCollection
+    ) -> Tuple[RateLimitState, KeyCollection, Sequence[Tuple[KeyResult, FetchType, FetchTypeExt]]]:
```

Comment:
A tuple containing a Tuple and Sequence of tuples should have a proper structured type. Otherwise it will be impossible to maintain.

Reply:
I moved the innermost tuple to a dataclass, lmk if that is sufficient